ehrql/query_model/nodes.py

```diff
@@ -87,22 +87,26 @@ class Node:
     "Abstract base class for all objects in the Query Model"
 
     def __init_subclass__(cls, **kwargs):
-        # This looks pointless, but `dataclass` overwrites inherited `__hash__` methods
-        # so we need this to be an explictly defined method
-        cls.__hash__ = cls.__hash__
         # All nodes in the query model are frozen dataclasses
         dataclasses.dataclass(cls, frozen=True)
+        # The __hash__ method gets clobbered by dataclasses, so we need to explicitly
+        # set it.
+        cls.__hash__ = Node.__hash__
 
     def __hash__(self):
         # Calculating the hash of an object requires recursively calculating the hashes
         # of its children. In a deeply nested query graph this can take some time and,
         # given how frequently `__hash__()` is called, this can end up completely
         # dominating ehrQL's execution time. Given that these are immutable
         # objects we can cache the hash value instead of recalcuting it each time.
+        # self._retrieved_from_cache is set to allow verification of hashing in tests.
         try:
-            return self.__hash__cache
+            value = self._hash_cache
+            object.__setattr__(self, "_retrieved_from_cache", True)
```

Comment:
I think there's a less intrusive way of testing this, which is to make use of the fact that `Value` nodes can take arbitrary objects. You can give it an object with some sort of spy attached to _its_ `__hash__` method and then confirm that the child hash method is called when expected.